jax/dtypes.py

```diff
@@ -105,14 +113,52 @@ def scalar_type_of(x):
   else:
     raise TypeError("Invalid scalar value {}".format(x))
 
+
+def _scalar_type_to_dtype(typ: type, value: Any = None):
+  """Return the numpy dtype for the given scalar type.
+
+  If `value` is specified and typ is int, then the resulting integer dtype will
+  be signed or unsigned depending on the value.
+
+  Examples
+  --------
+  >>> _scalar_dtype_to_dtype(int)
+  numpy.int64
+  >>> _scalar_dtype_to_dtype(float)
+  numpy.float64
+  >>> _scalar_dtype_to_dtype(complex)
+  numpy.complex128
+  >>> _scalar_dtype_to_dtype(int)
+  np.int64
+  >>> _scalar_dtype_to_dtype(int, 0)
+  np.int64
+  >>> _scalar_dtype_to_dtype(int, 1 << 63)
+  np.uint64
+  """
+  dtype = python_scalar_dtypes[typ]
+  # For integers, we return unsigned if required to store the value.
+  if typ is int and value is not None:
+    unsigned_dtype = _UINT_DTYPES[np.iinfo(dtype).bits]
```

Comment:
I'm surprised we don't consider x64 mode here. I guess this is intentional for now? If so leave a TODO.

Reply:
Yeah - the fact that we don't consider X64 mode in this utility I think has to do with this TODO: https://github.com/google/jax/blob/2947f562d53ae7d4f6ba9e87d9ccddca2f84a4d7/jax/dtypes.py#L56-L59

The reason I didn't canonicalize the type before checking the value is that I'm planning to address integer overflow in X32 mode separately (#5952). In fact, it was exploring that that led me to realize that the X64 mode behavior was inconsistent, so I decided to do this PR first.